File: src/lib/formatting.ts
/*
 * Text formatting, the way WhatsApp writes it (the app reads the same, in Formatting.kt):
 *   *bold*   _italic_   ~strikethrough~   ```monospace```   __underline__
 * A marker counts only at a word's edge, with no space just inside it, on one line. Markers can nest.
 * Pure, so the server (plain text for notifications) and the browser (the formatted message) share it.
 */

export type Mark = "mono" | "underline" | "bold" | "italic" | "strike";
export type Piece = string | { mark: Mark; children: Piece[] };

const MARKS: { token: string; mark: Mark }[] = [
  { token: "```", mark: "mono" },
  { token: "__", mark: "underline" },
  { token: "*", mark: "bold" },
  { token: "_", mark: "italic" },
  { token: "~", mark: "strike" },
];

const edge = (c: string | undefined) => c === undefined || !/[\p{L}\p{N}]/u.test(c);
const space = (c: string | undefined) => c !== undefined && /\s/.test(c);

function closing(text: string, open: number, token: string): number {
  const start = open + token.length;
  if (start >= text.length || space(text[start])) return -1;
  for (let j = start + 1; j <= text.length - token.length; j++) {
    if (text[j] === "\n") return -1;
    if (text.startsWith(token, j) && !space(text[j - 1]) && edge(text[j + token.length]) && !(token === "_" && text[j + 1] === "_")) return j;
  }
  return -1;
}

/** The text as plain pieces and marked groups. */
export function parseFormatting(text: string): Piece[] {
  const out: Piece[] = [];
  let plain = "";
  let i = 0;
  while (i < text.length) {
    const found = edge(text[i - 1]) ? MARKS.find((m) => text.startsWith(m.token, i)) : undefined;
    const close = found ? closing(text, i, found.token) : -1;
    if (found && close > 0) {
      if (plain) out.push(plain);
      plain = "";
      out.push({ mark: found.mark, children: parseFormatting(text.slice(i + found.token.length, close)) });
      i = close + found.token.length;
    } else {
      plain += text[i];
      i++;
    }
  }
  if (plain) out.push(plain);
  return out;
}

/** As [parseFormatting], but each marked group keeps its marker: what a text box shows while typing. */
export type LivePiece = string | { mark: Mark; token: string; children: LivePiece[] };

export function parseLive(text: string): LivePiece[] {
  const out: LivePiece[] = [];
  let plain = "";
  let i = 0;
  while (i < text.length) {
    const found = edge(text[i - 1]) ? MARKS.find((m) => text.startsWith(m.token, i)) : undefined;
    const close = found ? closing(text, i, found.token) : -1;
    if (found && close > 0) {
      if (plain) out.push(plain);
      plain = "";
      out.push({ mark: found.mark, token: found.token, children: parseLive(text.slice(i + found.token.length, close)) });
      i = close + found.token.length;
    } else {
      plain += text[i];
      i++;
    }
  }
  if (plain) out.push(plain);
  return out;
}

export const TOKEN: Record<Mark, string> = { bold: "*", italic: "_", underline: "__", strike: "~", mono: "```" };

/**
 * Put a marker round the selection, or take it off when it is already there (WhatsApp's menu). Spaces at the
 * selection's edges stay outside, since a marker only counts against a word. Gives the new text and selection.
 */
export function toggleMark(text: string, start: number, end: number, mark: Mark): { text: string; start: number; end: number } {
  const token = TOKEN[mark];
  let s = start;
  let e = end;
  while (s < e && /\s/.test(text[s])) s++;
  while (e > s && /\s/.test(text[e - 1])) e--;
  if (s === e) return { text, start, end };
  // Already marked, markers just outside the selection…
  if (text.slice(s - token.length, s) === token && text.slice(e, e + token.length) === token) {
    return { text: text.slice(0, s - token.length) + text.slice(s, e) + text.slice(e + token.length), start: s - token.length, end: e - token.length };
  }
  // …or just inside it.
  if (e - s > 2 * token.length && text.startsWith(token, s) && text.slice(e - token.length, e) === token) {
    return { text: text.slice(0, s) + text.slice(s + token.length, e - token.length) + text.slice(e), start: s, end: e - 2 * token.length };
  }
  return { text: text.slice(0, s) + token + text.slice(s, e) + token + text.slice(e), start: s + token.length, end: e + token.length };
}

/** The words alone, markers gone: notifications, previews, search. */
export function plainText(text: string): string {
  const flat = (pieces: Piece[]): string => pieces.map((p) => (typeof p === "string" ? p : flat(p.children))).join("");
  return flat(parseFormatting(text));
}

File: android/app/src/main/java/com/arkhins/wink/ui/components/Formatting.kt
package com.arkhins.wink.ui.components

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.background
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.TextRange
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.TextFieldValue
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.unit.dp
import com.arkhins.wink.ui.theme.Night
import com.arkhins.wink.ui.theme.Snow

/*
 * Text formatting, the way WhatsApp writes it, so people already know it:
 *   *bold*   _italic_   ~strikethrough~   __underline__ (WhatsApp has none; a double underscore here)
 * A marker counts only at a word's edge (so snake_case and a*b stay as they are), with no space just inside it, on
 * one line. Markers can nest: *_bold italic_*. The website reads the same (src/lib/formatting.ts).
 */

private enum class Mark(val token: String, val style: SpanStyle) {
    Underline("__", SpanStyle(textDecoration = TextDecoration.Underline)),
    Bold("*", SpanStyle(fontWeight = FontWeight.Bold)),
    Italic("_", SpanStyle(fontStyle = FontStyle.Italic)),
    Strike("~", SpanStyle(textDecoration = TextDecoration.LineThrough)),
}

private fun edge(c: Char?) = c == null || !c.isLetterOrDigit()

/** Where [mark] opened at [open] closes, or -1. */
private fun closing(text: String, open: Int, mark: Mark): Int {
    val start = open + mark.token.length
    if (start >= text.length || text[start].isWhitespace()) return -1
    var j = start + 1
    while (j <= text.length - mark.token.length) {
        if (text[j] == '\n') return -1
        if (text.startsWith(mark.token, j) && !text[j - 1].isWhitespace() && edge(text.getOrNull(j + mark.token.length)) &&
            // A single "_" isn't the start of a "__".
            !(mark == Mark.Italic && text.getOrNull(j + 1) == '_')
        ) return j
        j++
    }
    return -1
}

private fun AnnotatedString.Builder.appendFormatted(text: String) {
    var i = 0
    val plain = StringBuilder()
    while (i < text.length) {
        val mark = if (edge(text.getOrNull(i - 1))) Mark.entries.firstOrNull { text.startsWith(it.token, i) } else null
        val close = mark?.let { closing(text, i, it) } ?: -1
        if (mark != null && close > 0) {
            append(plain.toString())
            plain.clear()
            withStyle(mark.style) { appendFormatted(text.substring(i + mark.token.length, close)) }
            i = close + mark.token.length
        } else {
            plain.append(text[i])
            i++
        }
    }
    append(plain.toString())
}

/** A message's words with their formatting applied (the markers themselves hidden). */
fun formatted(text: String): AnnotatedString = buildAnnotatedString { appendFormatted(text) }

/** The words alone, markers gone: for previews, quotes and popups. */
fun plainText(text: String): String = formatted(text).text

/** The selection wrapped in [token] — or, when it already is, unwrapped. The selection stays on the same words. */
fun toggleMark(value: TextFieldValue, token: String): TextFieldValue {
    val text = value.text
    var from = value.selection.min
    var to = value.selection.max
    // A marker can't sit against a space: leave spaces at the selection's ends outside it.
    while (from < to && text[from].isWhitespace()) from++
    while (to > from && text[to - 1].isWhitespace()) to--
    if (from == to) return value
    val n = token.length
    return if (from >= n && text.startsWith(token, from - n) && text.startsWith(token, to)) {
        val out = text.substring(0, from - n) + text.substring(from, to) + text.substring(to + n)
        TextFieldValue(out, TextRange(from - n, to - n))
    } else {
        val out = text.substring(0, from) + token + text.substring(from, to) + token + text.substring(to)
        TextFieldValue(out, TextRange(from + n, to + n))
    }
}

/** B I U S over the message box while words are selected. */
@Composable
fun FormatBar(value: TextFieldValue, onChange: (TextFieldValue) -> Unit, modifier: Modifier = Modifier) {
    Row(modifier.padding(horizontal = 6.dp, vertical = 2.dp), horizontalArrangement = Arrangement.spacedBy(6.dp)) {
        listOf(
            Triple("B", "*", SpanStyle(fontWeight = FontWeight.Bold)),
            Triple("I", "_", SpanStyle(fontStyle = FontStyle.Italic)),
            Triple("U", "__", SpanStyle(textDecoration = TextDecoration.Underline)),
            Triple("S", "~", SpanStyle(textDecoration = TextDecoration.LineThrough)),
        ).forEach { (label, token, style) ->
            Box(
                Modifier
                    .size(36.dp)
                    .background(Night, RoundedCornerShape(10.dp))
                    .clickable { onChange(toggleMark(value, token)) },
                contentAlignment = Alignment.Center,
            ) {
                Text(buildAnnotatedString { withStyle(style) { append(label) } }, style = MaterialTheme.typography.titleMedium, color = Snow)
            }
        }
    }
}
